quark/subproject.py

```
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+        stamp_file = join(self.directory, ".stamp.quark")
+        if exists(join(stamp_file)):
+            with open(stamp_file, "r") as fileobj:
+                data = json.load(fileobj)
+                if data["job_id"] == self.stamp["job_id"] and data["url"] == self.stamp["url"]:
+                    return
+
+        # The stamp has changed and the directory already exists: we need to start from scratch.
+        if isdir(self.directory):
+            shutil.rmtree(self.directory)
+
+        with tempfile.TemporaryDirectory(prefix="quark-") as tempdir:
+            archive_path = self._download(tempdir)
+
+            if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz", ".zip")):
+                self._extract(archive_path, tempdir)
+            else:
+                os.mkdir(self.directory)
+                shutil.copy2(archive_path, self.directory)
+
+        # Update the stamp file
+        with open(stamp_file, "w") as fileobj:
+            json.dump(self.stamp, fileobj)
+
+    def url_from_directory(self, directory, include_commit=True):
+        stamp_file = join(directory, ".stamp.quark")
+        if not exists(stamp_file):
+            raise QuarkError("Missing stamp file: " + stamp_file)
+
+        with open(stamp_file, "r") as fileobj:
+            stamp = json.load(fileobj)
+
+        if not include_commit:
+            return stamp["url"]
+
+        if not stamp["sha1"]:
+            raise QuarkError("Missing sha1 in stamp file: " + stamp_file)
+
+        # Resolve job ID if needed.
+        url = urllib.parse.urlparse(stamp["url"])
+        fragments = Subproject._parse_fragment(url) if url.fragment else {}
+        if "ref" in fragments:
+            print_msg("resolving job id for " + self.parsed_artifact_name, self._comment())
+            stamp["job_id"] = self._resolve_job_id(
+                self.parsed_project_name,
+                fragments["ref"],
+                fragments["job"],
+            )
```

IIUC you are going to have a race condition here, as the job for the last pipeline for the given ref may be different from the one whose artifact was downloaded when doing `quark up`, so you end up with a bizarre URL containing sha1 and url that do not match the job_id.

---

Yup, it's kind of a bummer. Alternatively we can raise an error and force the caller to use a PAT to perform the initial download (in case they want to freeze later), so that we can have a stable job id at all times.